### ISSUE TITLE
Borked UA_Client lifecycle

### ISSUE THREAD

**Issue by Stasik0** [acplt/open62541#350]
Hi, 

The lifecycle of the client is borked as shown by problems by @tpiat e.g. #348

Normal client lifecycle is:
**LS 1**
* new
* connect
* *couple of reads*
* disconnect
* delete

new allocates only memory for the struct and does the init job

connect allocates memory 
* by calling a callback in the networklayer e.g. buffer - always
* allocates some memory in case of successful connection e.g. authentication token

or one from client.c exmaple
**LS 2**
* new
* connect
* --- server down --
* delete

**These two seem to be okay and tested**

More complex e.g. with error cases seem to be borked...

For example (without errors):
**LS 3**
* new
* connect
* *couple of reads*
* disconnect
* connect
* *couple of reads*
* disconnect
* delete

In this case disconnect needs to 'restore' the state of new somehow? And also to free memory.
***I added UA_Client_deleteMembers to UA_Client_delete to solve it, now we have a problem of double frees in #348 #349***

then there is the case:
**LS 4**
* new
* connect
* couple of reads
* --- server drops connection ----
* ??

do we need to call disconnect now? Is the connection closed e.g. buffer freed? If yes by whom? The error happens in the networklayer that does not have pointer to the client.

Currently everything is somehow hacked (*partially* messed up by my 'fixes') and complicated due to client-securechannel-connection interleaving and reusage of networklayer by server/client. I cannot fix it alone - need some design decisions:
* who frees and initializes the client? 
* what is the user expected to do in the error case?
* should disconnect 'invert' the application of connect? (in this case we need to deleteMembers of the client as i currently do)

cases we need to cover besides the ones above (the best way via unit-tests):

**LS 5**
* new
* connect
* *couple of reads*
* --- server drops connection ----
* ??
* connect
* disconnect 
* delete

and

**LS 4**
* new
* connect
* *couple of reads*
* --- server drops connection ----
* ??
* delete

and

**LS 6**
* new
* connect
* --- server down ----
* ??
* connect
* *couple of reads*
* disconnect
* delete

And then there are subscriptions...

**Comment by Stasik0** [comment]
I'd propose the following 

**(WI1)** _init and _deleteMembers functions for UA_client so it can reside on stack

Rule 1: **You have to call _disconnect if _connect returned success**
=> Implication 1 **(WI2)** _connect has to clean up in case it fails internally e.g. if the server is down
=> Consistent with **LS 1**

Rule 2: **_disconnect reverts the changes of _connect e.g. _disconnect = _deleteMembers + _init since _new does not allocate memory (we are lucky atm)**
=> **(WI3)** testing needed especially in subsequent _disconnect + _delete = _deleteMembers is called twice

**Comment by tpiat** [comment]
My sample code (localget, #348)

```
int localget()
{
	string uri("opc.tcp://");
	uri+=node.ip;

	if(!node.client_connected)
	{		
		UA_StatusCode retval = UA_Client_connect(node.client, ClientNetworkLayerTCP_connect,(char *)uri.c_str());
		if(retval != UA_STATUSCODE_GOOD)
		{

			UA_Client_disconnect(node.client);

			UA_Client_delete(node.client);
			node.client = UA_Client_new(UA_ClientConfig_standard, core_Logger);

			node.client_connected=0;
			return retval;
		}
		else node.client_connected=1;
	}

	UA_ReadResponse rResp = UA_Client_read(node.client, node.rReq);
	if(rResp.responseHeader.serviceResult != UA_STATUSCODE_GOOD)
	{
		UA_Client_disconnect(node.client);
		node.client_connected=0;		
	}
	UA_ReadResponse_deleteMembers(&rResp);

	return 0;
}
```
With disconnect or without it disconnect,
with Detete/new pair or witout it 
-- memory leak if server down.

versions 1cfd71af28 or b0e4f46304

**Comment by tpiat** [comment]
If it's intresting:

I added few log lines, and I see, where occur memory alloc or free
```
#if 0
#ifndef UA_free
# define UA_free(ptr) free(ptr)
#endif
#ifndef UA_malloc
# define UA_malloc(size) malloc(size)
#endif
#ifndef UA_calloc
# define UA_calloc(num, size) calloc(num, size)
#endif
#ifndef UA_realloc
# define UA_realloc(ptr, size) realloc(ptr, size)
#endif

#else
#include <stdio.h>
#define UA_free(ptr) UA_free1(__LINE__,ptr)
#define UA_malloc(size) UA_malloc1(__LINE__,size)
#define UA_realloc(ptr, size) UA_realloc1(__LINE__,ptr, size)
#define UA_calloc(num, size) UA_calloc1(__LINE__,num, size)

int mlc=0,fre=0;
int memdebug=1;

void *UA_malloc1 (int line,size_t size)
{
	void *ptr = malloc(size);
	if(memdebug)
	{
		mlc++;
		FILE *f=fopen("malloc.log","a+t");
		fprintf(f,"%p, line=%6d\n",ptr,line);
		fclose(f);
	}
	return ptr;
}

void *UA_realloc1(int line,void *block, size_t size)
{
	void *ptr = realloc(block,size);
	if(memdebug)
	{
		FILE *f=fopen("realloc.log","a+t");
		fprintf(f,"%p, line=%6d\n",ptr,line);
		fclose(f);
	}
	return ptr;
}

void UA_free1(int line,void *block)
{
	if(!block)return;

	if(memdebug)
	{
		fre++;
		FILE *f=fopen("free.log","a+t");
		fprintf(f,"%p, line=%6d\n",block,line);
		fclose(f);
	}
	free(block);
}

void *UA_calloc1 (int line,size_t num, size_t size)
{
	void *ptr = calloc(num,size);

	if(memdebug)
	{
		mlc++;
		FILE *f=fopen("malloc.log","a+t");
		fprintf(f,"%p, line=%6d\n",ptr,line);
		fclose(f);
	}
	return ptr;
}
#endif
```

**Comment by Stasik0** [comment]
@tpiat sorry for deleting your message, i will transfer it back to #348 

following idea, i have add ```client->status``` we get the following graph

![client_life](https://cloud.githubusercontent.com/assets/469170/9541830/1f23ca8e-4d6c-11e5-9e14-dc619fdbc101.png)

In the new branch https://github.com/acplt/open62541/compare/client_lifecycle I tried to implement it.

I also implemented some "fool-safe" functions for the users e.g. calling conenct multiple times will not cause a memleak, but will call a disconnect or a reset first. Is it okay? Should we habituate users with this?
